spec/acceptance/before_actions_spec.rb

```diff
@@ -0,0 +1,109 @@
+require 'spec_helper'
+require 'test_doubles'
+
+RSpec.describe 'Action before_actions' do
+  describe 'works with simple organizers - from outside' do
+    it 'can be used to inject code block before each action' do
+      TestDoubles::AdditionOrganizer.before_actions =
+        lambda do |ctx|
+          ctx.number -= 2 if ctx.current_action == TestDoubles::AddsThreeAction
+        end
+
+      result = TestDoubles::AdditionOrganizer.call(0)
+
+      expect(result.fetch(:number)).to eq(4)
+    end
+
+    it 'works with iterator' do
+      TestDoubles::TestIterate.before_actions = [
+        lambda do |ctx|
+          ctx.number -= 2 if ctx.current_action == TestDoubles::AddsOneAction
+        end
+      ]
+
+      result = TestDoubles::TestIterate.call(:number => 0,
+                                             :counters => [1, 2, 3, 4])
+
+      expect(result).to be_success
+      expect(result.number).to eq(-4)
+    end
+  end
+
+  describe 'can be added to organizers declaratively' do
+    module BeforeActions
+      class AdditionOrganizer
+        extend LightService::Organizer
+        before_actions (lambda do |ctx|
+                          if ctx.current_action == TestDoubles::AddsOneAction
+                            ctx.number -= 2
+                          end
+                        end),
+                       (lambda do |ctx|
+                          if ctx.current_action == TestDoubles::AddsThreeAction
+                            ctx.number -= 3
+                          end
+                        end)
+
+        def self.call(number)
+          with(:number => number).reduce(actions)
+        end
+
+        def self.actions
+          [
+            TestDoubles::AddsOneAction,
+            TestDoubles::AddsTwoAction,
+            TestDoubles::AddsThreeAction
+          ]
+        end
+      end
+    end
+
+    it 'accepts before_actions hook lambdas from organizer' do
+      result = BeforeActions::AdditionOrganizer.call(0)
+
+      expect(result.fetch(:number)).to eq(1)
+    end
+  end
+
+  describe 'works with callbacks' do
+    it 'can interact with actions from the outside' do
+      TestDoubles::TestWithCallback.before_actions = [
```

Comment:
I'm not sure why, but I'm having tingles about allowing people to add those callbacks dynamically at calling time on top of the other way, declared in the organizer. I'm thinking of people wondering "There's a thing going on, and I'm not sure what it is", and then going to their actions and organizer, not seeing it, and not getting where it can be coming from. Not sure if that's relevant.

Reply:
We need to be able to modify the organizer and inject logic from the outside. See [this PR](https://github.com/adomokos/light-service/pull/141). The [modifications](https://github.com/jeremy-hanna/light-service/blob/4381ad539c1c79f246990b6c52ef1943005500e7/lib/light-service/testing/context_factory.rb#L11-L17) to ContextFactory triggered this PR.

The great things about this is that it's there, you don't need to use it if you don't want to. And if you do, use it with great care. Sounds a lot like Ruby metaprogramming. 😉

Reply:
Sounds like you thought it through, no problem for me!